### DISCUSSION TITLE
PGN Viewer - Database file?

### DISCUSSION THREAD

**Discussion by sbernsto** [mliebelt/pgn-viewer#546] [General]
I am looking for a PGN viewer that is able to read a file that has chapters. What this looks like in an actual file is multiple PGNs saved in the same file.  I am trying to replicate Lichess Study viewer where you can have multiple chapters. Can this PGN-Viewer accomplish this?

**Comment by mliebelt** [comment]
Yes, of course, this is possible. The example https://mliebelt.github.io/pgn-viewer/examples/1101.html shows how to construct that. The one parameter you need is `manyGames: true`. If you want to read the games from a file, here is the docu https://mliebelt.github.io/pgn-viewer/docu.html with the parameter `pgnFile`, you should be able to read from file, not pgn as a string.

Good luck in trying out that part.

**Reply by sbernsto** [reply]
Thanks for your help! Two more question for you. 
1. If I had a number of positions in different games where there was a tactic, is your program able to display tactics puzzles? 
2. Is it possible to do some sort of interactive lesson like some lichess study's have where the user needs to make a move and not just walk through the pgn

**Reply by mliebelt** [reply]
A user implemented a mode named `pgnPuzzle` which may be sufficient for your needs. Here are the ideas how to get what you want:

1. Tactics puzzle: use `pgnPuzzle` with a start position denoted by `fen` parameter. Use then the main line as the solution (has to start with the party that has to solve the puzzle).
2. Interactive lesson: well, I don't understand what is the difference to the puzzle. A lesson should show comments, ... during the lesson, right? Could you elaborate what you mean by that?

**Reply by sbernsto** [reply]
![Screenshot 2024-05-30 at 1 03 09 PM](https://github.com/mliebelt/pgn-viewer/assets/6393618/b827880e-d9f5-41c8-82a5-672a6153ef6f)
see the screenshot attached.  I guess it is the same as a puzzle...

**Reply by mliebelt** [reply]
Yes, more or less. I don't see any difference in tutorial vs. puzzle. But the support is at the moment very limited by the `pgn-viewer`.

**Reply by sbernsto** [reply]
Thanks for your comments and assistance.  One final question for you.  I'm looking at analyzing a players last 20 games and extracting their worst 20 moves (not necessarily one from each game but worst 20 moves overall).  I was thinking I can then put them into some sort of puzzle so they can see what they did wrong and how to correct.  Do you think this would have to be 20 different files or could it be a single file with multiple pgns that pgnPuzzle could receive?  From our previous discussion and the documentation, pgnPuzzle is looking for a FEN and not a PGN.  I'd love your thoughts

**Reply by mliebelt** [reply]
Well, the  `pgnPuzzle` was developed by someone else, with minimal effort to have it. So yes, it may be that `pgnPuzzle` is not able to handle multiple games. Never thought about that, but it is  interesting, of course. Have you tried it? The viewer (better: the parser, that is used by the viewer) is able to read the current position from FEN, without anything about it in the configuration. See the combination of [SetUp and FEN in the specification](https://github.com/mliebelt/pgn-spec-commented/blob/main/pgn-specification.md#97-alternative-starting-positions). Then you could use the `pgnPuzzle` with that. Try it!

**Reply by sbernsto** [reply]
We've got PGN Viewer implemented and running.  Check out our beta site prior to production: https://beta.knightlychess.com/user/learn-chess/

Quick questions: I see two issues with a PGN.  First, any comment before the first move doesn't show up.  Second, if the comment has a ' such as I've or Black's the viewer fails.  I was curious if you've come across either of these two before?

Thanks!

Example PGN: 
[Event "London Opening: Introduction"]
[Site "https://lichess.org/study/pLBSiPGF/wdGlcwDu"]
[Result "*"]
[Variant "Standard"]
[ECO "D00"]
[Opening "Queen's Pawn Game: Accelerated London System"]
[Annotator "https://lichess.org/@/Sbernsto"]
[UTCDate "2024.05.20"]
[UTCTime "15:09:38"]

{ Hi! You've probably heard about an opening for white called the London System. In this study, I will teach you the basics of this popular opening. }
1. d4 { We will be starting with d4. It is not as aggressive as e4 can be but d4 is a good middleground between aggressive and solid games.
To start off as an example I will show you the ideas where black plays 1...d5. 1...Nf6 is a popular choice as well. } { [%cal Gd7d5,Gg8f6] } 1... d5 2. Bf4 { This is the trademark move of the London Opening, where you play d4 and Bf4. The idea is to build a solid pyramid in the center with e3 and c3, and the bishop on f4 will be very well placed.
This is as far as I will go for this chapter. In the next few chapters I will go into more detail on the ideas. } { [%csl Ge3,Gc3,Gf4][%cal Ge2e3,Gc2c3] } *

**Comment by mliebelt** [comment]
See recent issue https://github.com/mliebelt/pgn-viewer/issues/555 which relates to a discussion. So get notified on resolution of that bug then. Could you file a new issue about behavior where the reader fails due to not correct used quotes character inside comments? That should be handled  by the viewer of course, I am not sure where the bug comes in, parser or reader or viewer. But a new ticket that isolates that would help.
